src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                var arg2 = string.Empty;
+                if (parameter.ContainingSymbol.IsExternallyVisible() &&
+                    !parameter.ContainingSymbol.IsLocalFunction() &&
+                    !parameter.ContainingSymbol.IsAnonymousFunction())
+                {
+                    arg2 += FeaturesResources.if_it_is_not_part_of_a_shipped_public_API;
+                }
+
+                if (hasReference)
+                {
+                    arg2 += FeaturesResources.comma_its_initial_value_is_never_used;
+                }
+
+                var diagnostic = DiagnosticHelper.Create(s_unusedParameterRule,
+                                                         location,
+                                                         option.Notification.Severity,
+                                                         additionalLocations: null,
+                                                         properties: null,
+                                                         arg1,
+                                                         arg2);
+                reportDiagnostic(diagnostic);
+            }
+
+            private static IEnumerable<INamedTypeSymbol> GetAttributesForMethodsToIgnore(Compilation compilation)
+            {
+                // Ignore conditional methods (One conditional will often call another conditional method as its only use of a parameter)
+                var conditionalAttribte = compilation.ConditionalAttribute();
+                if (conditionalAttribte != null)
+                {
+                    yield return conditionalAttribte;
+                }
+
+                // Ignore methods with special serialization attributes (All serialization methods need to take 'StreamingContext')
+                var onDeserializingAttribute = compilation.OnDeserializingAttribute();
+                if (onDeserializingAttribute != null)
+                {
+                    yield return onDeserializingAttribute;
+                }
+
+                var onDeserializedAttribute = compilation.OnDeserializedAttribute();
+                if (onDeserializedAttribute != null)
+                {
+                    yield return onDeserializedAttribute;
+                }
+
+                var onSerializingAttribute = compilation.OnSerializingAttribute();
+                if (onSerializingAttribute != null)
+                {
+                    yield return onSerializingAttribute;
+                }
+
+                var onSerializedAttribute = compilation.OnSerializedAttribute();
+                if (onSerializedAttribute != null)
+                {
+                    yield return onSerializedAttribute;
+                }
+
+                // Don't flag obsolete methods.
+                var obsoleteAttribute = compilation.ObsoleteAttribute();
+                if (obsoleteAttribute != null)
+                {
+                    yield return obsoleteAttribute;
+                }
+            }
```

Comment:
as above, we should have an escape hatch that allows one to specify relevant metadata-names for attributes that should cause us to ignore.